txscript/pkscript.go

```diff
@@ -0,0 +1,194 @@
+// as a map index.
+type PkScript struct {
+	// class is the type of the script encoded within the byte array. This
+	// is used to determine the correct length of the script within the byte
+	// array.
+	class ScriptClass
+
+	// script is the script contained within a byte array. If the script is
+	// smaller than the length of the byte array, it will be padded with 0s
+	// at the end.
+	script [maxLen]byte
+}
+
+// ParsePkScript parses an output script into the PkScript struct.
+// ErrUnsupportedScriptType is returned when attempting to parse an unsupported
+// script type.
+func ParsePkScript(pkScript []byte) (PkScript, error) {
+	var outputScript PkScript
+	scriptClass, _, _, err := ExtractPkScriptAddrs(
+		pkScript, &chaincfg.MainNetParams,
+	)
+	if err != nil {
+		return outputScript, fmt.Errorf("unable to parse script type: "+
+			"%v", err)
+	}
+
+	if !isSupportedScriptType(scriptClass) {
+		return outputScript, ErrUnsupportedScriptType
+	}
+
+	outputScript.class = scriptClass
+	copy(outputScript.script[:], pkScript)
+
+	return outputScript, nil
+}
+
+// isSupportedScriptType determines whether the script type is supported by the
+// PkScript struct.
+func isSupportedScriptType(class ScriptClass) bool {
+	switch class {
+	case WitnessV0PubKeyHashTy, ScriptHashTy, WitnessV0ScriptHashTy:
+		return true
+	default:
+		return false
+	}
+}
+
+// Class returns the script type.
+func (s PkScript) Class() ScriptClass {
+	return s.class
+}
+
+// Script returns the script as a byte slice without any padding.
+func (s PkScript) Script() []byte {
+	switch s.class {
+	case WitnessV0PubKeyHashTy:
+		return s.script[:witnessV0PubKeyHashLen]
+	case ScriptHashTy:
+		return s.script[:scriptHashLen]
+	case WitnessV0ScriptHashTy:
+		return s.script[:witnessV0ScriptHashLen]
+	default:
+		// Unsupported script type.
+		return nil
+	}
+}
+
+// Address encodes the script into an address for the given chain.
+func (s PkScript) Address(chainParams *chaincfg.Params) (btcutil.Address, error) {
+	_, addrs, _, err := ExtractPkScriptAddrs(s.Script(), chainParams)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse address: %v", err)
+	}
+
+	return addrs[0], nil
+}
+
+// String returns a hex-encoded string representation of the script.
+func (s PkScript) String() string {
+	return fmt.Sprintf("%x", s.Script())
+}
+
+// ComputePkScript computes the pkScript of an transaction output by looking at
+// the transaction input's signature script or witness.
+//
+// NOTE: Only P2SH, P2WSH, and P2WPKH redeem scripts are supported.
+func ComputePkScript(sigScript []byte, witness wire.TxWitness) (PkScript, error) {
+	var pkScript PkScript
+
+	// If a signature script is provided, since we only support P2SH for
+	// non-segwit scripts, then we'll assume this is a P2SH redeem script
+	// based on its length not being that of a P2PKH redeem script.
+	if len(sigScript) > 0 && len(sigScript) != pubKeyHashSigScriptLen {
```

Comment:
We can make this a bit stricter, as we _know_ that the sigScript for a witness program _must_ be a canonical push of the witness program and nothing more. We can borrow some logic from this section of the main VM code: https://github.com/btcsuite/btcd/blob/master/txscript/engine.go#L965

Reply:
If we intend to support regular P2SH, then we can't make this assumption as the redeem script can be arbitrary.